harmony/harmony.py

```diff
@@ -968,6 +985,124 @@ def download_all(self,
+            working directory by default.
+            overwrite: If True, will overwrite a local file that shares a filename with the
+            downloaded file. Defaults to False. If you're seeing malformed data or truncated
+            files from incomplete downloads, set overwrite to True.
+
+        Returns:
+            An Iterator that can be used to iterate over the granule results from a job
+        """
+        GET_JSON_RETRY_LIMIT = int(os.getenv('GET_JSON_RETRY_LIMIT', 3))
+        GET_JSON_RETRY_SLEEP = float(os.getenv('GET_JSON_RETRY_SLEEP', 1.0))
+        next_url = self._status_url(job_id)
+
+        # index used to keep track of where we are in the data links if the page gets reloaded
+        # so we don't pull the same granule more than once
+        # NOTE: this assumes that granule order is preserved when the page is reloaded and
+        # new granules are added. If this is not true then this won't work.
+        current_page_granule_count = 0
+
+        while next_url:
+            self_url = next_url
+            last_pull_time = datetime.now()
+            response = None
+            get_json_try_count = 0
+            # work around for occasional failures in the status page
+            while not response:
+                try:
+                    response = self._get_json(next_url)
+                except BaseException:
+                    response = None
+                    get_json_try_count += 1
+                    if get_json_try_count < GET_JSON_RETRY_LIMIT:
+                        time.sleep(GET_JSON_RETRY_SLEEP)
+                    else:
+                        raise Exception('Failed to get or parse job status page')
+
+            next_url = None
+            status = response.get('status')
+            if status == 'failed':
+                raise Exception(response.get('message'))
+
+            links = response.get('links', [])
+            link_index = 0
+            for link in links:
+                if link['rel'] == 'data':
+                    link_index += 1
+                    if link_index > current_page_granule_count:
+                        current_page_granule_count = link_index
+                        future = self.download(link['href'], directory, overwrite)
+                        bbox = link['bbox']
+                        temporal = link['temporal']
+                        yield {
+                            'path': future,
+                            'bbox': BBox(bbox[0], bbox[1], bbox[2], bbox[3]),
+                            'temporal': temporal
+                        }
+                elif link['rel'] == 'next':
+                    next_url = link['href']
+
+            if not next_url:
+                # no 'next' link means either we are done or we need to reload this page to
+                # get more results
+                if status in {'successful', 'paused', 'canceled', 'complete_with_errors'}:
```

Comment:
Shouldn't we also stop on 'failed' status? I see a test for failed in the tests, so I'm not sure how that works - does 'failed' status cause some exception to happen and exit this code? Might be worth a comment on why 'failed' is not in this list.

Reply:
We raise an exception when the job fails. Adding a comment as you requested.